pkg/ha/state/state.go

```diff
@@ -27,58 +45,116 @@ type StateView struct {
 	RecentLeaderWriteTime time.Time
 }
 
-func (s StateView) String() string {
-	return fmt.Sprintf("{\n\tLead:%s;\n\tUntil:%s;\n\tMaxSeen:%s;\n\tMaxInst:%s;\n\tMaxSeenLeader:%s\n}",
-		s.Leader, s.LeaseUntil, s.MaxTimeSeen, s.MaxTimeInstance, s.MaxTimeSeenLeader)
-}
+// Creates a new Lease and immediately synchronizes with the database, it either
+//	- sets the potentialLeader as the leader for the cluster with a lease
+//	  for the requested minT and maxT
+//	- or the existing leader and lease details are returned and set in the
+//	  new Lease.
+// An error is returned if an error occurred querying the database.
+func NewLease(c client.LeaseClient, cluster, potentialLeader string, minT, maxT, currentTime time.Time) (*Lease, error) {
+	dbState, err := c.UpdateLease(context.Background(), cluster, potentialLeader, minT, maxT)
+	if err != nil {
+		return nil, fmt.Errorf("could not create new lease: %#v", err)
+	}
+	var maxTimeInstance string
+	if maxT.After(dbState.LeaseUntil) {
+		maxTimeInstance = potentialLeader
+	} else {
+		maxTimeInstance = dbState.Leader
+	}
 
-// haLockState represents the current lock holder
-// as reported from the db
-type HALockState struct {
-	Cluster    string
-	Leader     string
-	LeaseStart time.Time
-	LeaseUntil time.Time
+	return &Lease{
+		cluster:               dbState.Cluster,
+		leader:                dbState.Leader,
+		leaseStart:            dbState.LeaseStart,
+		leaseUntil:            dbState.LeaseUntil,
+		maxTimeSeen:           maxT,
+		maxTimeInstance:       maxTimeInstance,
+		maxTimeSeenLeader:     maxT,
+		recentLeaderWriteTime: currentTime,
+	}, nil
 }
 
-func (h *State) UpdateStateFromDB(latestState *HALockState) {
+// UpdateFromDB uses the supplied client to attempt to update the lease for the potentialLeader.
+// It either updates the lease, or a new leader with the assigned lease interval is set, as
+// signified by the database as the source of truth.
+func (h *Lease) UpdateFromDB(c client.LeaseClient, potentialLeader string, minT, maxT time.Time) error {
+	h._mu.RLock()
+	cluster := h.cluster
+	oldLeader := h.leader
+	h._mu.RUnlock()
+	stateFromDB, err := c.UpdateLease(context.Background(), cluster, potentialLeader, minT, maxT)
+	if err != nil {
+		return fmt.Errorf("could not update lease from db: %#v", err)
+	}
```

Comment:
why not just call SetUpdateFromDB here?
```suggestion
	}
	h.SetUpdateFromDB(stateFromDB)
	return nil
```

Reply:
because of the explicit call from outside, i needed a check for cluster!=h.cluster, no longer needed